src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,241 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "mode-base-cluster-objects.h"
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributePersistenceProvider.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/af.h>
+#include <set>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeBase {
+
+class Delegate;
+
+class Instance : public CommandHandlerInterface, public AttributeAccessInterface
+{
+private:
+    Delegate * mDelegate;
+
+    EndpointId mEndpointId{};
+    ClusterId mClusterId{};
+
+    // Attribute data store
+    uint8_t mCurrentMode;
+    DataModel::Nullable<uint8_t> mStartUpMode;
+    DataModel::Nullable<uint8_t> mOnMode;
+    uint32_t mFeature;
+
+    // CommandHandlerInterface
+    void InvokeCommand(HandlerContext & ctx) override;
+    CHIP_ERROR EnumerateAcceptedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context) override;
+    CHIP_ERROR EnumerateGeneratedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context) override;
+
+    // AttributeAccessInterface
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+
+    /**
+     * Internal change-to-mode command handler function.
+     */
+    void HandleChangeToMode(HandlerContext & ctx, const Commands::ChangeToMode::DecodableType & req);
+
+    /**
+     * Helper function that loads all the persistent attributes from the KVS. These attributes are CurrentMode,
+     * StartUpMode and OnMode.
+     */
+    void LoadPersistentAttributes();
+
+    /**
+     * Helper function that encodes the supported modes.
+     * @param encoder The encoder to encode the supported modes into.
+     */
+    CHIP_ERROR EncodeSupportedModes(const AttributeValueEncoder::ListEncodeHelper & encoder);
+
+public:
+    /**
+     * Initialise the ModeBase server instance.
+     * @return Returns an error if the cluster ID given is not af a valid ModeBase cluster, if the
+     * given endpoint and cluster ID have not been enabled in zap, if the CommandHandler or
```

Comment:
```suggestion
     * @return Returns an error if the given endpoint and cluster ID have not been enabled in zap, if the CommandHandler or
```
because we took the "valid alias" bit out, no?